### ISSUE TITLE
rowdetails or nested grids

### ISSUE THREAD

**Issue by maxim-s-barabash** [6pac/SlickGrid#128]
Hi team,
I need to Grid can show additional information below each grid row.
What can you advise? thank.

RowDetails.
![rowdeteil](https://user-images.githubusercontent.com/2724325/27985475-20dd5aac-63f4-11e7-9900-7fd1912ff12f.png)
or Nested Grids
![nested](https://user-images.githubusercontent.com/2724325/27985476-24ee70ea-63f4-11e7-9f94-8043e9c6ebde.png)

like http://www.jqwidgets.com/jquery-widgets-demo/demos/jqxgrid/rowdetails.htm

**Comment by 6pac** [comment]
Not easily. Slickgrid rows are all fixed height. This is not changeable.
I do have a SlickCombo control that drops down a cell into a separate grid. But that's my own private control, and it wasn't that easy to create, and I'm not sure it's fully debugged yet.

**Comment by maxim-s-barabash** [comment]
> Slickgrid rows are all fixed height

Now I'm working on changing this, without losing compatibility. I plan to use getItemMetadata for this.
I hope for this will need a minimum of changes.

**Comment by 6pac** [comment]
The height calcs are buit into the grid. Imagine if there are a million rows. To work out the grid rendering, the grid has to work out what row we are at, the size of the next x rows and add them.

Ag-grid, probably the best alternative to Slickgrid, has the option for non-fixed-height rows, but only if you switch off the virtual scrolling feature. Slickgrid's virtual scrolling is built in at a low level.

What you *may* be able to do is assign several rows to create an aggregated row that is a multiple of the standard row height, then use a scroll bar. Even that's going to be complicated.

**Comment by ghiscoding** [comment]
I'm also looking into this functionality and I was happy to fall on this [Stack Overflow - Can SlickGrid's row height be dynamically altered?](https://stackoverflow.com/questions/10535164/can-slickgrids-row-height-be-dynamically-altered#29399927) question. That person who answered on the question also posted a [JsFiddle](http://jsfiddle.net/violet313/98wx61uz/) example and she also has a lengthy [article](http://violet313.org/slickgrids/) about it. I tried the JsFiddle and modified it to work with the row detail and it works, it's quite cool too, so we could easily add a template to make it easier. 

I'm totally going to look into in the coming weeks, her answer was posted in April 2015 and so I'm not sure how different is the grid code since then but it shouldn't be that bad and she said it's 200 line of JS code. If I ever have the time (or anyone of you have the time) to add an example demonstrating this, it would be very cool.

**Comment by SatanEnglish** [comment]
@ghiscoding 
Have you had a chance to look into this yet?
I found that post myself a while back still haven't found any chance either to try and see if it works.

My current project has been using 6pac new SlickGrid and I'm hoping not to have to move away from SlickGrid as I really like using it.

**Comment by ghiscoding** [comment]
Unfortunately no, I'm more busy trying to wrap up all features of 6pac SlickGrid into an [Angular-SlickGrid](https://github.com/ghiscoding/Angular-Slickgrid) and [Aurelia-SlickGrid](https://github.com/ghiscoding/aurelia-slickgrid), I'd really like to try out this functionality but that won't be until I got all features wrapped up first. 

However I'm contributing when I can to 6pac version, since this is the best fork available. I guess that the ideal way of implementing this would be to create a Plugins or a Control so that it would be optional and not part of the core itself. Though I haven't tested anything, so I don't even know if this is doable without core changes. 

Anyone tried it?

**Comment by SatanEnglish** [comment]
@ghiscoding Ok wow I put the 200lines of code in.
Works like a charm I'm only using it as a hole pane across the rows.
Checked the reordering via a header and the data ends up in the right place. (If pane is closed)
Only down fall with the 200 line one is that you would have to close the expanded panes when reordering is used.
Just working through the blog to see what's involved to fix the reordering while pane is open.

PS: No change in the core was needed here!

**Comment by ghiscoding** [comment]
@SatanEnglish I would be more interested in her first version, that is with a detail pane without any column shown in it, not sure if that is the 200 lines you said. Anyhow for the reordering, I would simply use a [Grid Event](https://github.com/mleibman/SlickGrid/wiki/Grid-Events) like `onColumnsReordered` would possibly work to just close the pane on reorder trigger. 

It would be nice if you can make it a SlickGrid plugin (with demo) that anyone of us could use just by calling a SlickGrid register plugin

**Comment by SatanEnglish** [comment]
@ghiscoding 

Here is a very simplistic Plugin version incase I never getting round to finishing it.
It is basiclly the detail pane from the example as a plugin. (using the checkbox plugin example)
It is currently just putting in the random data like the example.

Js Plugin
[slick.detailsView.txt](https://github.com/6pac/SlickGrid/files/1709485/slick.detailsView.txt)
css
[slick.detailView.css.txt](https://github.com/6pac/SlickGrid/files/1709535/slick.detailView.css.txt)

`
var detailSelectColumn= new Slick.DetailSelectColumn({
  cssClass: "detailView-toggle"
});
columns.unshift(detailSelectColumn.getColumnDefinition());

grid.registerPlugin(detailSelectColumn);
`

**Comment by ghiscoding** [comment]
Thanks @SatanEnglish I decided to take a look at your first draft of the plugin. It wasn't working at first, I had to do a couple of changes for it to work (mainly 2 things, it was missing `_dataView = _grid.getData();` in the `init()` function and `getPaddingItem()` function was missing as well). Also note that the plugin only works with a DataView since that it what is used in the plugin itself. 

After that was resolved, I copied some code with a DataView and made a few styling modifications and finally replaced the text from `oook` (which didn't mean much) to instead show a template of the actual current item. The result is show in the print screen below and I also create fork and branch to this out, you can try from this [Github branch](https://github.com/ghiscoding/SlickGrid/tree/feature/row-detail) and here's the [commit](https://github.com/ghiscoding/SlickGrid/commit/64882b7637153650d3519fdea159207e2b0da763) of the first working draft.

Feel free to make changes on the branch and/or create your own PR. I was just curious to try it out with the code you gave us :)
I think another cool idea would be to possibly pass a template as an option to the plugin. Then the plugin would have some kind of ways to parse the template to replace item info with actual data. I'm not sure how that would work though since the actual item data is from the grid and I would rather pass a lot of more item detail. Anyhow, if we could pass a template to the plugin, as well as detail object and possible some kind of formatter (or parser) to display the row detail. 


![slickgrid-row-detail](https://user-images.githubusercontent.com/643976/36012471-4990df82-0d2c-11e8-984a-8504dd7fe0d9.png)

**Comment by SatanEnglish** [comment]
@ghiscoding

My apologizes I had updated the JS file to include the getPaddingItem(), But you must of grabed it inbetween my update.

Am I right in thinking that you have also fixed the height in your version to load from the height of the content?

A template would be nice. (Not useful for loads of data)

An ajax call that loads in a partial is what i'm needing. (And a spinner while it loads)
Also I saw something in that post about caching data if the ones that are loaded via ajax could be stored so they don't need reloading that would make it fast and useful

**Comment by ghiscoding** [comment]
I found the missing function in the original article describing the code. I copied it and it started working. Then I modified your css to add some icons for expanding/collapsing.

The height of the panel is dependent on the line height and adding a variable to hold the number of rows your pane has seems to be the key. In my case there's 6 lines of detail, so that seems to be the number to use. 

Yeah I would also use it with async (ajax) calls as well since I would use the id to call for full row detail. 

So, I was thinking of maybe using `asyncPostRender`, like this [example](http://6pac.github.io/SlickGrid/examples/example10-async-post-render.html) since this is similar to an async formatter. I also think creating some kind of a callback Formatter is the key for creating custom template to provide to the plugin. I'm not too worried about caching or anything, I think displaying a spinner while it's loading would be decent enough.

**Comment by SatanEnglish** [comment]
@ghiscoding

Never used the asyncPostRender so not sure how well I'll manage it.
I'll have a bit of a play around with it if I can.

Formatter wise would you see this as something you would add to the generic Formatters in the long run?

Or would it just be another file to include like the plugin and css in the case you want to use the plugin?

@6pac Input on some of this would be nice to as he's maintaining the main Repo

**Comment by 6pac** [comment]
This is an interesting innovation to the grid. If it can be delivered as a plugin, then @ghiscoding has commit, and i'm happy for you to workshop it and commit the result, or issue a PR.
If it turns out to be too hard to do in a plugin, then I'd have to look at it to see to what degree it complicates the base code.
As usual, I'm really busy, and PRs/issues are backing up a bit. I'm in the process of putting together a small Slickgrid website to help co-ordinate new features and changes, and also to try to attract some of the traffic which currently goes to the MLeibman repo.

**Comment by ghiscoding** [comment]
I gave another go at it, and got a first working proof of concept with a delay. I was totally off the field with using `asyncPostRender`, it has nothing to do with this since the post renderer will render every single row in a delayed fashion which is not what we are looking for. 

Instead everything is in the plugin and how it will handle things. The first working concept is with a manual `setTimeout` to add a 1 sec. delay in displaying something to the user (which is a real life concept). To the plugin itself, I added a few options, first of is a `preTemplate`, that acts as a pre-loader if we may say (so in other word, we can display a spinner or a "loading" message in that template). Once the process is over (the delay), it will call another property for the `postTemplate` which has the item as argument (`item` will be replaced with your custom object in the future). I also need to use another property called `templateRowCount` to know how many rows (lines) does the final template has, this might be automated in the future but for now it will suffice for it to work. 

A piece of the code
```javascript
function loadingTemplate() {
      return '<div><h4>Loading...</h4></div>';
    }

    function postTemplate(item) {
      return filledTemplateRows = [
        '<div>',
        '<h4>' + item.title + '</h4>',
        '<div class="detail"><label>Duration:</label> <span>' + item.duration + '</span></div>',
        '<div class="detail"><label>% Complete:</label> <span>' + item.percentComplete + '</span></div>',
        '<div class="detail"><label>Start:</label> <span>' + item.start + '</span></div>',
        '<div class="detail"><label>Finish:</label> <span>' + item.finish + '</span></div>',
        '<div class="detail"><label>Effort Driven:</label> <span>' + item.effortDriven + '</span></div>',
        '</div>'
      ].join('');
    }

// create the row detail plugin
      var detailSelectColumn = new Slick.DetailSelectColumn({
        cssClass: "detailView-toggle",
        preTemplate: loadingTemplate,
        postTemplate: postTemplate,
        templateRowCount: 6
      });
```

The next step would be to replace the `setTimeout` by a `Promise` or something else depending on what you guys think is the best to use. I assume that I will have another property possibly called `process` or `processing` which will contain the async handling with a `Promise` or whatever else.

You can check where I added a `setTimeout` in the [plugin](https://github.com/ghiscoding/SlickGrid/blob/a8f89de2f1e0dfdd5b79f5afa8a561fd78ec8636/plugins/slick.detailview.js#L106) 

And here's an animated gif of the proof of concept. I first open "Task 2" and 1 sec. later you can see the row detail of it. 


![2018-02-12_01-14-55](https://user-images.githubusercontent.com/643976/36085468-fa7da5b4-0f93-11e8-9d7e-a776399b1d75.gif)

**Comment by ghiscoding** [comment]
I had a thought not long after I went to bed, I keep forgetting that SlickGrid is amazing and has everything it needs directly in it. So what we need to complete the code is simply to create a `Slick.Event` inside the plugin which we will subscribe directly in the plugin since the notify will be called by the client side code and that's it we are in business :)

So now the client side code calls a `notify` once the row detail item comes back like this
```javascript
function simulateServerCall(item) {
      // fill the template on delay
      setTimeout(function() {
        detailSelectColumn.onBeforeRowDetailOpen.notify({
          "detailItem": item
        }, undefined, this);
      }, 1000);      
    }

// the plugin creation has a new `process` property
detailSelectColumn = new Slick.DetailSelectColumn({
        cssClass: "detailView-toggle",
        preTemplate: loadingTemplate,
        postTemplate: postTemplate,
        process: simulateServerCall,
        templateRowCount: 6
});
```

and inside the plugin we suscribe to the event
```javascript
// subscribe to the onBeforeRowDetailOpen so that the plugin knows when the user server side calls finishes with it's data back
_self.onBeforeRowDetailOpen.subscribe(function (e, args) {
      item._detailContent = _options.postTemplate(rowDetailItem);
      var idxParent = dataView.getIdxById(rowDetailItem.id);
      dataView.updateItem(rowDetailItem.id, rowDetailItem);
});
```

works like a charm 💯

**Comment by SatanEnglish** [comment]
@ghiscoding

Unfortunately I've been ill this week so I haven't had a chance look at any of this yet. (But looks like u have got a good version going)

Having an overflow sounds smart to me I'd assume this would just make the area scroll-able if there is to much content.

Still wondering if we can get the load to put the view in a hidden get the height and then make the height right size for the view but that will have to be a later addition if it's do-able.

I should be back to it tomorrow then I should get time to try it out and give you some feedback.

I'm actually going to be trying to load tabs in here which will all likely have different size data so I really am going to have to use MAGIC here and remove and add padding on the fly as the tabs are changed.

Thanks for finding time for this.

PS: as you say "SlickGrid is amazing" This is why I'm trying to help implement this so I don't have to go to another grid.

**Comment by ghiscoding** [comment]
@SatanEnglish  oh sorry to hear you got ill, hope you'll be back in shape in no time. 

So yes I'd like to work on the overflow first and see if that is doable. Not exactly sure what you mean by 

> put the view in a hidden get the height and then make the height right size for the view

Did you mean to simply calculate the height of the view pane and keep it hidden until the row detail is ready? If that's the case then I guess I could do that, what I have now is that it does the calculation from the `preTemplate` (where it shows "Loading...") and so I do have and know the height from the start. If that is what you are looking for, I could add a flag for that (since some user might want to show the pane with a loading, while others might want to only show the pane only when the row detail is ready)

Feel free to do any commit to my [fork](https://github.com/ghiscoding/SlickGrid), it all started from your code and you will most probably be the first one to use it :)

**Comment by SatanEnglish** [comment]
@ghiscoding

I have actually checked the ordering of the columns and the ordering asyc and desc with my orginal plugin and it worked. (Closing the detailView when it does)

I'll double check with your changes that it still works tomorrow.

Nice going on the overflow!

Estimated height by picking a row amount sounds ok to me.

Did you see any issue with the view when half if off the top of the screen?

I notice in the version in the blog that if half is on the screen the image disapears. I'm hoping this is only with images and not going to affect the detailView. (Again I'll have a look once I get a chance but worth noting incase you hadn't noticed in the blog)

Example is below images below.

![image](https://user-images.githubusercontent.com/2100563/36136131-cb500492-10f3-11e8-8900-f8324939e57b.png)

![image](https://user-images.githubusercontent.com/2100563/36136122-bf4dbab8-10f3-11e8-9641-34d4742a601e.png)

**Comment by ghiscoding** [comment]
@SatanEnglish 

Since I took your plugin, it does have the closing on orderBy then, which is great thanks. I just need to enable sorting in my example to test it out. 

Didn't know about the visual glitch you just mentioned, I haven't read the original article in a while but yes I see the effect on my side too. It might be caused by the buffering (top/bottom) that SlickGrid does. If I scroll back to see the entire row detail, it all show up again, so yeah a little annoying but not a deal breaker. 

Can you try using my version and see how it goes on your side. We're getting closer to a first PR with this, you can also take the code and create a PR yourself if you want it associated to your name (I really don't mind, it all started because of you which is great)

**Comment by SatanEnglish** [comment]
@ghiscoding 

Can you give me rights to push to your branch?
Just checking stuff now and found a small amount of missing stuff.

I can always just type it here but if it's possible to let me commit to your branch that would work to!

**Comment by ghiscoding** [comment]
I added you and 6pac as collaborator to the [fork](https://github.com/ghiscoding/SlickGrid/tree/feature/row-detail), again make sure to use the [feature/row-detail](https://github.com/ghiscoding/SlickGrid/tree/feature/row-detail) branch. Feel free to do any changes you want, just get it done :)

**Comment by SatanEnglish** [comment]
@ghiscoding

git -c diff.mnemonicprefix=false -c core.quotepath=false push -v --tags origin feature/row-detail:feature/row-detail
remote: Permission to ghiscoding/SlickGrid.git denied to SatanEnglish.
fatal: unable to access 'https://github.com/ghiscoding/SlickGrid.git/': The requested URL returned error: 403

I'll try again later maybe permissions just not filtered through yet.

Otherwise I'll just add files here if I change them.

**Comment by ghiscoding** [comment]
@SatanEnglish  Kinda new to that too, so let me know if that doesn't work.

**Comment by SatanEnglish** [comment]
@ghiscoding 

I've added some functionality to the detailView.
You can now directly load in a page/View.

e.g. 
rowdetailView.onAsyncResponse.notify({
                    "itemDetail": item,
                    "detailView": "<div><h4>It's a View</h4></div>"
}, undefined, this);

I've also added a bool for _options.loadOnce
Which tell it if it should reload the view or use whats already loaded.

I haven't set up an example of this yet as you may wish to have a look and do some tweaking.

I've check that your example still works and I have an example where I'm getting some view from a server using mine.

PS: Passing and object from the server to your object for the template worked by the way.

**Comment by 6pac** [comment]
PS. I am watching all this unfold, but haven't looks closely at any code yet. Let me know when it reaches a stable point and I'll review. I suppose that will be in the form of a PR :-)  
If it's a plugin or has only minor mods to core code, I'm happy to pull the beta and let the refinements come in via successive PRs. That's not going to destabilise any other features.

**Comment by SatanEnglish** [comment]
@6pac 

Currently there are **0** changes to any core stuff it is literally just the plugin and a css file so far.

**Comment by SatanEnglish** [comment]
@ghiscoding

Got an idea for how to deal with the height better would like some input when you have a minute.

Basic example would be to have a div with id = temp (Not actually thinking to call it temp)

Css would be as follows

`<div id="temp" style="position:absolute; left:9999999px; top:9999999px;">``</div>`

script
// Load the view into the tempView
$("#temp").html(view);
// Get the Height of the view
 var height = document.getElementById('temp').clientHeight;

Now we have the actual height of the of the view!

We would of course then have to have a maxHeight so that it can be contained better.
May have to have some calculation that works out the closest row amount. (Else may break the rows)

Thoughts?

**Comment by SatanEnglish** [comment]
@6pac @ghiscoding
Is there a way to change/override the "getRenderedRange" function without changing core.js?

Current issue lets say I set the panelRow for how big the detailview is to 10 rows.
When the first row goes off the screen we lose our detailView as they are being cleaned up.

The function cleanupRows(rangeToKeep) in slick.grid.js is removing the row because it's off the screen.

Ideally I want to override it so that we keep X rows off the top were X is the 10 rows specified above.
The bottom already holds extra rows from what I can tell but not the top.

**Comment by 6pac** [comment]
Not possible to say for me without reviewing the code carefully. If you can make small edits to the core (eg. pass an extra param to ```getRenderedRange``` or ```cleanupRows``` to trigger different behaviour, I'm good with that.  I will go over this in detail sometime soon.

**Comment by SatanEnglish** [comment]
@6pac 

I've cheated.

Feel free to make it better in the long run with passing an extra param.

init method added => _grid.getRenderedRange = getRenderedRange;

Then I have cloned the getRenderedRange from core and changed
                var minBuffer = 3;
To
                var minBuffer = _options.panelRows + 3;

Nasty but all contained in the plugin WOOT!

**Comment by ghiscoding** [comment]
@SatanEnglish yeah I thought the row buffer was the cause of the visual glitch, I think that as @6pac said, it wouldn't be that bad to be able to pass an argument to change the `minBuffer` without affecting the current default when nothing is set. I'm sure this could be useful for not just this plugin but other things too. If that's the only change in the core, it's not too bad. I always try to avoid touching core as much as possible but in some cases it's just better to go with a change.

**Comment by SatanEnglish** [comment]
@ghiscoding 

I've added an example for the changes that I added. [example17-row-detail-view-load.html](https://github.com/ghiscoding/SlickGrid/blob/feature/row-detail/examples/example17-row-detail-view-load.html)

This has both the view per row functionality and the onetime loading.

Can you have a fiddle and if you don't see any issues I'm happy for you or me to do the first PR.

Feel free to clean it up a little if you notice any untidy bits.

Please have a fiddle and see if anything breaks!!

**Comment by ghiscoding** [comment]
@SatanEnglish it's nice to have flexibility I guess, so it's great to support both with/without template. In my code.

We need to fix the sorting before creating a PR, not sure if you saw my previous comment about the `handleSort` throwing an error after I added Sorting in the example16. Could you take a look at it?

**Comment by SatanEnglish** [comment]
@ghiscoding
Assuming your still doing some doing stuff here.
I've got a magical wrapper that exposes the rows...

The right one for the generic stuff should be the following.

    function handleSort(e, args) {		
      var rows = _grid.getData().getItems();		
      $.each(rows, function (i, r) {
          HandleAccordionShowHide(r, true);
      });
    }
Feel free to fix this if your going to test this now.

**Comment by ghiscoding** [comment]
@SatanEnglish yes it does fix it however this is extremely slow, it takes like 20sec to do a sort. If I comment out the handleSort in the plugin then it goes back to normal which is sorting in millisecond. That doesn't seem ideal... can we do it on just the rows in the buffer instead of the entire dataset?

**Comment by SatanEnglish** [comment]
@ghiscoding

No sure what the issue is my version is really fast however I am using a far better sort than the default.
Also I already have the rows in my version.
I'll double check where i'm getting my rows from when I get to work tomorrow..

But an array sounds find. I'll just have to check the onTimeload version still works after a sort. (Or set the item to not being loaded again)

**Comment by ghiscoding** [comment]
I made a commit with the sort and the `handleSort` fix last night and so you can try it from the example16. However there's huge performance hit looping through the entire dataset. I really think that having an array of the open ones is the way to go. You could also make it a public function `collapseAll` that would loop through this array of references, it should be instant to do so.

**Comment by SatanEnglish** [comment]
@ghiscoding

I've check this and didn't **SEE** a performance issue but I can try implementing a holder of the set

**Comment by ghiscoding** [comment]
would still prefer to use an array of open references, this is by far the most optimized way of dealing with open/collapse of row details, especially with huge dataset. You could probably see a difference if you put in 500k rows in the dataset (just increase the for loop number in the example16). Your computer is probably faster than mine then, since it really was 20sec. on my side versus half a second when commenting out the `handleSort` in the plugin.

**Comment by SatanEnglish** [comment]
@ghiscoding

Done and Done.

- Running of an array
- Added a collapseAll
- Separated into more function for more readability

**Comment by SatanEnglish** [comment]
@ghiscoding @6pac

You guys interested in an AutoColumnSizePlugin?

**Comment by ghiscoding** [comment]
@SatanEnglish 
A quick look at it and it seems all nice and sweet. I'll review it more later tonight after work. I didn't even know javascript `filter` function was supported in IE9+ without a polyfill (sweet). Ahh wait it seems we forgot 1 small thing in the name of the plugins, it should be under the namespace "Slick.Plugins.pluginName" (for example [HeaderMenu Plugin](https://github.com/ghiscoding/SlickGrid/blob/79ebf31121769bc4436bb6dc5035edac0c73eda5/plugins/slick.headermenu.js#L5) and ours is under "Slick.pluginName", if you can fix it that would be great, else I can fix it later tonight. 

As for the `AutoColumnSizePlugin`, I tried the example from this [demo page](https://cdn.rawgit.com/naresh-n/slickgrid-column-data-autosize/master/examples/example-1.html) and I'm very interested. Can you make a separate PR for it and I can review it as well. 

@6pac, for the PRs that are related to plugins, do you still want to review them all or do you want me to merge them right away (I'm talking about the RowDetailPlugin and possibly `AutoColumnSizePlugin`)

**Comment by SatanEnglish** [comment]
@ghiscoding
Changed to Slick.Plugins.pluginName

After you look at it tonight I'm happy with the plugin now so feel free to get @6pac   to commit it

**Comment by SatanEnglish** [comment]
@ghiscoding

As for the AutoColumnSizePlugin I'll make an example and have a fiddle after work.
Any chance you could set up the other PR for it I actually not 100% sure how to do it.

**Comment by ghiscoding** [comment]
@SatanEnglish to do a PR is easy when you follow these steps. Do a fork from Github and do a checkout locally, work on the code then push (commit) the code to your Github fork, from the Github page click on the Pull Request button (as shown below). You can all do this on a master branch (from your fork) or what is best is to actually create a different branch so that you won't mix your code with other features if you have multiple PR open.

Here's the PR button
![pr](https://user-images.githubusercontent.com/643976/36231320-9bac87c0-11ab-11e8-9405-7bc20ff137f7.png)

Give that a try, it's nice to see our name in the contributor list of an active project :)

**Comment by SatanEnglish** [comment]
@ghiscoding 
Another thing to look into with the DataViewRow.
I'm wanting to have an option to have a row tap so that tapping on any item will open the view.

Not sure how I'd go about this any ideas?

**Comment by 6pac** [comment]
@ghiscoding feel free to merge this PR if you like. I've made some small updates to the repo and been halfway through a few extra things for a while, so haven't got around to doing a release yet.   
I think it's time to release though, so when you're done, let me know.

**Comment by ghiscoding** [comment]
@SatanEnglish 
I've reviewed the code
- added a `useRowClick` as per your request
- added a few Slick Events that user can subscribe to
- added lots of documentation (on top of the plugin)

Here's a couple of things before officially opening the PR
- can you review the docs (on top of the plugin) and see if I missed anything
- would it be possible to combine example16 and example17 into 1 html file?
  - we could have 2 grids into the same page
- and lastly what happened with this comment you wrote earlier?
> In particular updating the minBuffer seems to fix this issue.
var range = getVisibleRange(viewportTop, viewportLeft); var buffer = Math.round(viewportH / (options.rowHeight * 10)) ; var minBuffer = 3;
Aprox line 1914 in slick.grid.js

**Comment by SatanEnglish** [comment]
@ghiscoding 

1) Missing something about being able to use the args.detailView if wanting to add an explicit view for the onAsyncEndUpdate. (Not sure if this is done on purpose if not can u add it )

2) Making making exmaples into one sounds fine is there a way we can add some buttons that toggle on "loadOnce", "loading by template vs loading by view", "Row Tap on/off" 

3) I have overrideden the core by changing the function in the plugin e.g.

`_grid.getRenderedRange = getRenderedRange;` line 79 of plugin

**Comment by SatanEnglish** [comment]
@ghiscoding @6pac 

Hay guys really quick example setup.
[Example Auto-Resize-Columns](https://github.com/SatanEnglish/SlickGrid/tree/Aucto-Resize-Columns)

Seems to not work perfectly if using double click on column resize now. (Think there is some change to the nest version that has affected it)

Resize all button works like a charm tho.

Think the issue line in plugin below as I think I read something in the updates that these bits are now editable/variables on grid?
`clone.style.cssText = 'position: absolute; visibility: hidden;right: auto;text-overflow: initial;white-space: nowrap;';`

Anyway have a play.

**Comment by ghiscoding** [comment]
@SatanEnglish 
Not sure what you mean by 
> 1. Missing something about being able to use the args.detailView if wanting to add an explicit view for the onAsyncEndUpdate. (Not sure if this is done on purpose if not can u add it )

I added doc and also a console.log example in example16 to demonstrate all 3 new onEvents, isn't that sufficient? or did you see some bugs?

2. yeah I guess we could try with a radio button to active the onLoad, but I'm not sure if we can change the plugin instance after the grid already loaded the plugin. We would have to try, else we'll have to make 2 grids. 

For your last comment about the other plugin, I can look at that later but we should try to keep the discussion here about the row detail plugin. I suggest you to open another Github issue about that plugin and we can discuss it there.

**Comment by SatanEnglish** [comment]
@ghiscoding

1) This is what I mean.

        onAsyncEndUpdate: Fired when the async response finished
        Event args:
        grid:        Reference to the grid.
        itemDetail:  Column definition.
        detailView:  An explicit view to use instead of template (Optional)

2) Two grids is fine if thats what we need to resort to!

**PS** I've started deleting comments that are are off topic just leaving the latest ones. (Planning to add another issue some stage)

**Comment by SatanEnglish** [comment]
@ghiscoding

Small change that you may want to add.
Currently the hover for the row click doesn't change to the hand icon should it?
e.g. like it does if you hover over the icon column

**Comment by ghiscoding** [comment]
@SatanEnglish 

1. I already had the docs for the `onAsyncEndUpdate`, you can see this [line](https://github.com/ghiscoding/SlickGrid/blob/7e65176f19adfad9c2cb3023ceb348135a420459/plugins/slick.rowdetailview.js#L26)

2. haven't tried 2 grids or just a radio button yet, will do that soon

3. having a hand cursor when hovering the row... hmm really? 

Did you push your changes? I don't see any changes on the branch since my yesterday's code.

**Comment by SatanEnglish** [comment]
@ghiscoding 

1) Updated the missing comment now it's on onAsyncResponse not the end my bad.

3) I've looked at how your doing it I think skip putting a nice hover on... will just have to stay not looking like you can click on the row.

**Comment by ghiscoding** [comment]
@SatanEnglish 

- I managed to merge both examples into `example16.html`, I didn't delete the 17 yet for references. 
  - Can you please review 16 and make sure it's up to your needs. I don't fully understand your Load Once, so maybe provide better info on the home page?
  - Please note that I changed the multiple of 2 for a multiple of 5 because it includes both examples and more views. You can change it to another multiple if you.
- I really don't know how to add the hand cursor when the `loadOnce` is defined by javascript. So I'll skip that part. 
  - However, if you are using that option on all your grids, then just change the CSS and add `cursor: pointer` to `slick-row`. That is the most effective way of doing it, since it would be much slower to deal with that in javascript. 

So I ask you 1 final review and once you confirm, I will create the official PR.

**Comment by SatanEnglish** [comment]
@ghiscoding 

Ok yer you didn't seem to understand how this worked at all LOL.

I've updated the example now have another look and read the descriptions.
If you still don't understand how it works please let me know.

If you cannot understand how these work it's not a good enough description.

Can you revise again please.

**Comment by ghiscoding** [comment]
PR is in the queue, see below for an animated of the result and multiple features that were added on the way. Thanks for @SatanEnglish for the awesome help on this team work effort :)

Totally understand the "loadOnce" now, totally awesome! No more waiting when re-opening the same row :)

Also note that the option of "useRowClick" will not add the hand cursor over the entire row, you have to add the proper CSS yourself if you intend to always use that option. That is not a bug, it's simply expensive performance wise to do that in javascript while a simple CSS (if you always use that option) will be super fast.

![2018-02-16_12-50-57](https://user-images.githubusercontent.com/643976/36321588-48a32566-1318-11e8-8f4a-3e3c3948f7a8.gif)

**Comment by ghiscoding** [comment]
@SatanEnglish 
I noticed a small issue with your "loadOnce" implementation. It looks like it's only keeping the last loaded one in memory and if you had multiple ones that got opened, they will all get the last loaded one showing. For example, take this way to reproduce, Open task 2 (loading+template), Open task 4 (loading+template), close/open task 4 (from cache), close/open task 2 (from cache but shows task 4 not 2). 

I didn't check the code but I assume you keep last loaded somewhere in cache, but you might have to keep them in an array instead of just loading last one.

**Comment by SatanEnglish** [comment]
@ghiscoding 

Hmm I hadn't noticed but that's likely because we have so little difference between items.

Can you have a look at line 194 of the plug in.

I'm basically just taking the it directly off the item which I thought was the row item.
e.g. If we already have item._detailContent on the item we are looking at then I just use it as is.

Alternatively I'll have to have a look at this when I get to implementing on the current project which may not be for a week or so.

**Comment by SatanEnglish** [comment]
@ghiscoding

Seems the override of getRenderedRange isn't working anymore either for some reason

**Comment by ghiscoding** [comment]
I don't think that I've touched that function neither do I know how you implemented the loadOnce. Sorry I can't help you much on that

**Comment by SatanEnglish** [comment]
Yer I'm looking into some of this now

**Comment by SatanEnglish** [comment]
@ghiscoding
It's not the onetime load thats broken it's the hole thing!!

Basically as soon as you scroll the view off the screen and then back on it's wrong!!
Doesn't seem to matter if it's the one time load or not!!

**Comment by SatanEnglish** [comment]
@ghiscoding 
I've found the issue not 100% sure how to fix it!!
_self.onAsyncResponse.subscribe(function (e, args) {
This function is called for every item in the grid that is currently open

I have updated some stuff here firstly I have added a return if the item being looked at is not our item
if(item.id != args.id) return;
Is there a better way to deal with this in the notify explicitly? 

Secondly just incase we have to leave it the way it is I have updated everything to run off the "args.itemDetail" instead of updating the item.

Can you please have a look at this.
Firstly is there a way we can fix notify to only notify the one in question?
Secondly we may want to change the other bits I've changed back to item not the args item

**Comment by SatanEnglish** [comment]
@ghiscoding @6pac 

I'm still looking into updating the getRenderedRange.
But it looks like I may need to update the slick.grid.js to have a variable after all

**Comment by 6pac** [comment]
not a problem if you need to do that.

**Comment by SatanEnglish** [comment]
@6pac 
Doesn't look like I can get around changing slick.grid.js.
But it will be a minimal change. Going to add "adjustRowBuffer: 0" to the var defaults.

Then just have   
          `var minBuffer = 3 + options.adjustRowBuffer;`

**Comment by SatanEnglish** [comment]
@6pac 
Actually i'm just going to add minRowBuffer as a hole to the options and default it to the 3.
I'm sure you will change it if you don't like it and let me know.

**Comment by SatanEnglish** [comment]
@6pac 
Has this branch been merged in yet?
If so can you at your next available chance re-merge this as I have at least fix the main issues.
Double check the small change to slick.grid.js.

Thanks

**Comment by ghiscoding** [comment]
@SatanEnglish 
The PR has not been merged yet, however all changes applied on the same branch are part of the PR. So your last 3 commits are part of the PR. 

We, or I always try to avoid touching the core but, If you're still fighting for 1 line of code to change in the core, please let us know and I'm sure @6pac can get it done if necessary. Also note that the branch created for this PR is most probably outdated with current master branch, if you need a change to the core, it's better to have a separate PR for that (even if it's just 1 line of code).

**Comment by SatanEnglish** [comment]
@ghiscoding 
I've changed it in one of the commits.
Just the thing around seeing if you can make your notify calls a little more efficient.(I'm not sure how that part works.)

But all works right again now

**Comment by ghiscoding** [comment]
@SatanEnglish  
The notify are just something that gets called when it's reached. Not sure which of the notify you meant, but if you want it to trigger less often, because of a condition, then just wrap it in an `if` condition.

**Comment by SatanEnglish** [comment]
@ghiscoding 

Inside expandItem(item)

_self.onAsyncResponse.subscribe(function (e, args) {      

Not sure what `if` I would wrap it in.
Its currenly being call for every open item item.

When you have some time run it and put a break point inside the subscribe.
Then open row 1. (it will be hit once)
Then open row 3. (it will be hit twice)
Then open row 7. (3 times)

Not sure if we need to be unsubscribing or something here but seems something isn't quiet right.

**Comment by ghiscoding** [comment]
@SatanEnglish 
good catch, it was being called exponentially simply because the `expandItem` was adding an extra subscriber every time the function was being called, which was making it exponential. However now that the subscriber is outside of the `expandItem` and the fact that your last commit added a comparison with the `item.id`, I had to add a reference to the [_currentItem](https://github.com/ghiscoding/SlickGrid/blob/3d211d2e96bf48ee40bbbdb226cd8ae0ea643cd7/plugins/slick.rowdetailview.js#L180) inside the `expandItem` and use that for the comparison inside the `onAsyncResponse` as can be seen on line [212](https://github.com/ghiscoding/SlickGrid/blob/3d211d2e96bf48ee40bbbdb226cd8ae0ea643cd7/plugins/slick.rowdetailview.js#L212)

Please re-test your stuff, my quick tests seems to be all good on from what I can see.

**Comment by SatanEnglish** [comment]
@ghiscoding 
I'm going to have a look at testing it now.

Can you also look at 
html.push("<div class='detail-container' style='max-height:"

I'm thinking that this should be height instead of max height so the content always takes up the full space thoughts? (Doesn't seem to affect your scroll if I do it.)

**Comment by SatanEnglish** [comment]
@ghiscoding

Now that you have fixed the onAynsResponse I have removed the `_currentItem`
Double check it now only fires on the item that is selected not all open items and therefore we no longer need to check the args vs current. (They are always the same object now)

Please double check again. (I did some testing here and seems to work)

**Comment by ghiscoding** [comment]
@SatanEnglish 
The difference with CSS `height` is that the `height` will fix it to that defined height, however a `max-height` will make it possible to go up to that maximum but never go above. Basically, the `max-height` is dynamic but never goes above. So, let say you define a `max-height: 300px` and your content is only 100px high, then it will have it's height showing as 100px, while a `height` is always set to that height no matter what. 

With that saying, why would you want to always show a full height instead of it being dynamic?

**Comment by SatanEnglish** [comment]
@ghiscoding

![image](https://user-images.githubusercontent.com/2100563/36363296-bddbab2a-15a0-11e8-8a82-1786860bdd89.png)

Don't as me Y the is currently a horizontal scroll but there is. But because it's max-height not height you can see the scroll bar is in a non sensible place

**Comment by ghiscoding** [comment]
@SatanEnglish 
If `height` fixes your problem then go ahead, I just thought dynamic height might be better but I haven't tried testing anything else than the demo yet so you're more advanced on that front.

**Comment by SatanEnglish** [comment]
@ghiscoding dynamic would be good if our rows were dynamic height all round.

**Comment by SatanEnglish** [comment]
@ghiscoding

Hay I don't know I want to commit this but have a play with this version
[slick.rowdetailview.js.txt](https://github.com/6pac/SlickGrid/files/1735941/slick.rowdetailview.js.txt)

I've hard coded the minBufferRows to 30 as I'm not sure what else we would do with it.

Try it with our example.

Dynamic size magic.

**Comment by ghiscoding** [comment]
@SatanEnglish 
You know more than I do on the `minBufferRows` since I don't have any demo sample other than the example, so I let you decide if it should be in the PR or not. On another note, I thought you would change the `max-height` to simply `height`, if it's better with `height` in your demo, please change it.

Could you also please finish your review on this PR and tell us when it's all good, it's not merged yet and it's starting to hold up release of a new SlickGrid version which I'm waiting for. We can't test and fix everything in 1 day (or 1 week), and if we need fixes in the future, we'll do other PR. Thanks

**Comment by SatanEnglish** [comment]
@ghiscoding
I'll double check once I get to my desk. 

Can u have a look at the version I sent in last Comment please check it for speed!

It's sizing to the content dynamically. (actual height)

**Comment by SatanEnglish** [comment]
@ghiscoding 
Have changed the max-height to height and have committed it.
I will keep looking into this dynamic height and add an option if I can get it working!

That way you can go ahead with the PR. As as is works fine for the first version

**Comment by ghiscoding** [comment]
@SatanEnglish 
Adding more features to this PR should be done in future version(s), unless it's real quick and can be done today. Basically, I'd like this PR to merge today, meaning in the next few hours. I think the current PR is good as it is and we can ship it and get feedback from other as we get more users trying it and developing with it.

I'm patiently waiting for a new version of SlickGrid :)

**Comment by SatanEnglish** [comment]
@ghiscoding
In case you want to have a play with the dynamic height version.

[slick.rowdetailview.js.txt](https://github.com/6pac/SlickGrid/files/1738360/slick.rowdetailview.js.txt)
This is were I got to it's still missing some scenario's but i'm going to go back to the version I know works.

I am however going to try and get a resizeDetailView function in before the release assuming your still waiting on the new SlickGrid version

**Comment by ghiscoding** [comment]
@SatanEnglish 
I am able to merge my own PRs, I was going to do that tonight assuming we had a stable version of our plugin. From what I know, our PR is holding up the release of a new SlickGrid version which is why I wanted this shipped. I didn't want to delay too long since I know it's holding the release. How long will it be for that work that you're doing on the resizeDetailView function?

**Comment by SatanEnglish** [comment]
@ghiscoding 

I've added a resize method to the code.
It's not called so it won't break anything for the release.

An example of using it is Open item task 2.
Then in console `detailView.resizeDetailView(_dataView.getItemById("id_2"))`

**Comment by SatanEnglish** [comment]
@ghiscoding 

.dynamic-cell-detail
What is the reason for using `display: table;`
I've came across case were I have a slick grid inside a slick grid and it breaks the width.

Is there a specific reason for having it?

**Comment by ghiscoding** [comment]
@SatanEnglish 
1. I think the `display: table` comes from the article. If you follow her article until the end, she wanted to an extra row of data underneath the row, which is different from our approach of a row detail. It's probably ok to remove it. I tested it on my side and can change the `display: table` to `display: inline` and the other one below from `display: table-cell` to `display: inline-block`
2. if you find that some CSS could/should be customized by the user, then please move it into the `html` file, outside of the plugin css. That way, we suggest something to the user without forcing them to use that.
3. I'm busy with my own repos releases tonight, so I won't have time to merge this PR tonight, will probably do it tomorrow.

**Comment by SatanEnglish** [comment]
@ghiscoding

Ok I think I have removed the main ones that aren't needed.

Hopefully this is all that's need now

**Comment by SatanEnglish** [comment]
@ghiscoding, @6pac 

Ok come across a snags for what I'm doing. (Not needed for first version)

I need to know when the columns are being re-aranged so that I can save the current state of the view.
(onColumnsReordered is after the view is already lost)

Basiclly lets say my text on my view is X and at somepoint it is changed to Y.
Then I reorder the columns by drag and drop.
The text in the view is returned to X as this is what it was at time of creation.

I already have a method for updating the view but not sure what to hook into.

**Comment by 6pac** [comment]
Hey @ghiscoding sorry about no update yet, not actually the PR, just me not having hours in the day. I'm happy to pull whatever is relatively stable and then monitor subsequent updates.

**Comment by ghiscoding** [comment]
@6pac it's ready to merge as it is, we'll do refinement in further versions. I was just too busy with my own merge yesterday.

**Comment by ghiscoding** [comment]
The merge is done, we should expect a new version of SlickGrid coming out soon. If you find any problems with the new plugin, please fill in new issue(s). Thank you.

**Comment by 6pac** [comment]
Hey @ghiscoding, can I ask what tool or process you use to generate your explanatory GIFs? They are no less than awesome.

**Comment by ghiscoding** [comment]
@6pac 
Yeah I found that software couple months ago, I always find animated GIF are perfect to explain/show something in a Github issue. The software is called [ShareX](https://getsharex.com/) and it's free, just click on the link. In my case, I installed it from the App Store in Windows 10. Free is awesome!

It has many type of output, here's the one I choose for the animated GIF. :)


![image](https://user-images.githubusercontent.com/643976/36633817-84737e04-1969-11e8-8160-d2fe77d2076f.png)
